src/app/public/modules/dock/dock.component.ts

```diff
@@ -0,0 +1,127 @@
+import {
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  ComponentFactoryResolver,
+  ElementRef,
+  Injector,
+  OnInit,
+  Type,
+  ViewChild,
+  ViewContainerRef
+} from '@angular/core';
+
+import {
+  SkyDockDomAdapterService
+} from './dock-dom-adapter.service';
+
+import {
+  SkyDockItemConfig
+} from './dock-item-config';
+
+import {
+  SkyDockItemReference
+} from './dock-item-reference';
+
+import {
+  sortByStackOrder
+} from './sort-by-stack-order';
+
+/**
+ * @internal
+ */
+@Component({
+  selector: 'sky-dock',
+  templateUrl: './dock.component.html',
+  styleUrls: ['./dock.component.scss'],
+  providers: [
+    SkyDockDomAdapterService
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class SkyDockComponent implements OnInit {
+
+  /**
+   * Use `any` for backwards-compatibility with Angular 4-7.
+   * See: https://github.com/angular/angular/issues/30654
+   */
+  @ViewChild('target', {
+    read: ViewContainerRef,
+    static: true
+  } as any)
+  private target: ViewContainerRef;
+
+  private itemRefs: SkyDockItemReference<any>[] = [];
+
+  constructor(
+    private changeDetector: ChangeDetectorRef,
+    private resolver: ComponentFactoryResolver,
+    private elementRef: ElementRef,
+    private injector: Injector,
+    private domAdapter: SkyDockDomAdapterService
+  ) { }
+
+  public ngOnInit(): void {
+    this.domAdapter.watchDomChanges(this.elementRef);
+  }
+
+  public insertComponent<T>(
+    component: Type<T>,
+    config: SkyDockItemConfig = {}
+  ): SkyDockItemReference<T> {
+
+    const factory = this.resolver.resolveComponentFactory(component);
+    const injector = Injector.create({
+      providers: config.providers || [],
+      parent: this.injector
+    });
+
+    const componentRef = this.target.createComponent(factory, undefined, injector);
+    const stackOrder = (config.stackOrder !== null && config.stackOrder !== undefined)
+      ? config.stackOrder
+      : this.getHighestStackOrder();
+
+    this.itemRefs.push({
+      componentRef,
+      stackOrder
+    });
+
+    this.sortItemsByStackOrder();
```

Comment:
I noticed new items with the same stack order get put underneath their matching stack order siblings, instead of on top. Is that deliberate?

Reply:
Yep! By default, an array of dock items that do not specify a `stackOrder` should match the "stack order" of the initializing array. For example:
```
private configs: any[] = [
  {
    backgroundColor: 'red'
  },
  {
    backgroundColor: 'white'
  },
  {
    backgroundColor: 'blue'
  }
];
```
Should produce:
![Screen Shot 2020-02-11 at 5 40 50 PM](https://user-images.githubusercontent.com/12497062/74286405-b1171580-4cf5-11ea-9a9a-e6d89c3d74aa.png)

Because of this requirement, adding individual items of the same `stackOrder` are added at the end of items with the same stack order.

Reply:
Awesome. Thanks for the explanation.